avm/src/vm/opcode.rs

```diff
@@ -2758,48 +2758,77 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
     Box::pin(async move {
-      let res = __httpget(HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0]))).await;
-      hand_mem.init_fractal(args[2]);
-      hand_mem.push_fixed(args[2], if res.is_ok() { 1i64 } else { 0i64 });
-      match res {
-        Ok(res) => hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal(&res)),
-        Err(ee) => hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal(format!("{}", ee).as_str())),
+      let req = hand_mem.read_fractal(args[0]);
+      let method = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&req, 0).0);
+      let url = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&req, 1).0);
+      let headers = hand_mem.read_from_fractal(&req, 2).0;
+      let mut out_headers = Vec::new();
+      for i in 0..out_headers.len() {
+        let header = hand_mem.read_from_fractal(&headers, i).0;
+        let key = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&header, 0).0);
+        let val = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&header, 1).0);
+        out_headers.push((key, val));
       }
-      hand_mem
-    })
-  });
-
-  async fn __httppost(uri: String, payload: String) -> Result<String, Box<dyn std::error::Error>> {
-    let url = TryInto::<Uri>::try_into(uri)?;
-    let request = Request::post(url).body(Body::from(payload))?;
-    let client = Client::builder().build::<_, Body>(HttpsConnector::new());
-    let mut response = client.request(request).await?;
-    let body = hyper::body::to_bytes(response.body_mut()).await?;
-    let res = String::from_utf8(body.to_vec())?;
-    Ok(res)
-  }
-  io!(httppost => fn(args, mut hand_mem) {
-    Box::pin(async move {
-      let uri = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0]));
-      let payload = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1]));
-      let res = __httppost(uri, payload).await;
+      let body = HandlerMemory::fractal_to_string(hand_mem.read_from_fractal(&req, 3).0);
+      let out_body = if body.len() > 0 { Some(body) /* once told me... */ } else { None };
+      let res = __httpreq(method, url, out_headers, out_body).await;
       hand_mem.init_fractal(args[2]);
       hand_mem.push_fixed(args[2], if res.is_ok() { 1i64 } else { 0i64 });
       match res {
-        Ok(res) => hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal(&res)),
-        Err(ee) => hand_mem.push_fractal(args[2], HandlerMemory::str_to_fractal(format!("{}", ee).as_str())),
+        Ok(mut res) => {
+          let mut res_hm = HandlerMemory::new(None, 3);
+          res_hm.init_fractal(0);
+          res_hm.push_fixed(0, res.status().as_u16() as i64);
+          let headers = res.headers();
+          let mut headers_hm = HandlerMemory::new(None, headers.len() as i64);
+          let mut i = 0;
+          for (key, val) in headers.iter() {
+            let key_str = key.as_str();
+            let val_str = val.to_str().unwrap();
```

Comment:
should we let the user handle invalid response headers?

Reply:
We should definitely guard against maliciously crafted headers instead of crashing here, I agree with you on that. The code is getting very branchy and nested as I tackle this, though.